### PR TITLE
Add stubs in IR and ABI for `f16` and `f128`

### PR DESCRIPTION
**Pull request by tgross35**
This is the very first step toward the changes in https://github.com/rust-lang/rust/pull/114607 and the [`f16` and `f128` RFC](https://rust-lang.github.io/rfcs/3453-f16-and-f128.html). It adds the types to `rustc_type_ir::FloatTy` and `rustc_abi::Primitive`, and just propagates those out as `unimplemented!` stubs where necessary.

These types do not parse yet so there is no feature gate, and it should be okay to use `unimplemented!`.

The next steps will probably be AST support with parsing and the feature gate.

r? @compiler-errors
cc @Nilstrieb suggested breaking the PR up in https://github.com/rust-lang/rust/pull/120645#issuecomment-1925900572

### CODE REVIEW

File: src/tools/clippy/clippy_utils/src/consts.rs [hunk 1]
```diff
@@ -778,8 +778,10 @@ pub fn mir_to_const<'tcx>(lcx: &LateContext<'tcx>, result: mir::Const<'tcx>) ->
                 let range = alloc_range(offset + size * idx, size);
                 let val = alloc.read_scalar(&lcx.tcx, range, /* read_provenance */ false).ok()?;
                 res.push(match flt {
+                    FloatTy::F16 => unimplemented!("f16_f128"),
```
Comment (flip1995) [inline src/tools/clippy/clippy_utils/src/consts.rs:781]: I'm a bit wary about adding this, especially to tools. There's definitely a risk of forgetting some of them, causing ICEs.

Comment (tgross35) [reply src/tools/clippy/clippy_utils/src/consts.rs:781]: I can change this, what would be better?

My original change did more https://github.com/rust-lang/rust/pull/114607/files#diff-83cfffe9dbc2758ecf7b34a433525befc32ebff4525aef0de105fa5e7b659c21R808-R815, but since the surface area of this PR is intentionally much smaller, the primitives aren't available for use.

File: tests/ui/parser/f16-f128.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Make sure we don't ICE while incrementally adding f16 and f128 support
```
Comment (Nilstrieb) [inline tests/ui/parser/f16-f128.rs:1]: I don't think parser is quite right, the parser doesn't care :)

File: src/tools/clippy/clippy_utils/src/consts.rs (hunk 1 above)
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/consts.rs:781]: I see the benefit of small PRs and this comment isn't meant to block the PR. Just a concern, that we shouldn't forget about those `unimplemented!` macros.

Comment (Nilstrieb) [reply src/tools/clippy/clippy_utils/src/consts.rs:781]: Removing the panics could be added as a step to the tracking issue

File: compiler/rustc_smir/src/rustc_internal/internal.rs [hunk 3]
```diff
@@ -166,8 +166,10 @@ impl RustcInternal for FloatTy {
 
     fn internal<'tcx>(&self, _tables: &mut Tables<'_>, _tcx: TyCtxt<'tcx>) -> Self::T<'tcx> {
         match self {
+            FloatTy::F16 => rustc_ty::FloatTy::F16,
```
Comment (compiler-errors) [inline compiler/rustc_smir/src/rustc_internal/internal.rs:169]: I'd rather leave f16/f128 out of stable mir altogether, since it's not stable at all. Just make it ICE if you try to convert it.

File: compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs [hunk 4]
```diff
@@ -470,8 +470,10 @@ fn encode_ty<'tcx>(
         // (See https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#fixed-width-floating-point-types.)
         ty::Float(float_ty) => {
             typeid.push(match float_ty {
+                FloatTy::F16 => unimplemented!("f16_f128"),
                 FloatTy::F32 => 'f',
                 FloatTy::F64 => 'd',
+                FloatTy::F128 => unimplemented!("f16_f128"),
```
Comment (compiler-errors) [inline compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:476]: @rcvalle: are there encodings for these types?

File: src/tools/clippy/clippy_utils/src/consts.rs (hunk 1 above)
Comment (compiler-errors) [reply src/tools/clippy/clippy_utils/src/consts.rs:781]: In fairness to @tgross35, this is impossible to hit right now. I think this approach is the best for the moment.

File: tests/ui/parser/f16-f128.rs (hunk 2 above)
Comment (compiler-errors) [reply tests/ui/parser/f16-f128.rs:1]: Yeah, this test does literally nothing lol

**Review by compiler-errors — APPROVED** [review]
Some tiny nits, then I think this is fine

File: tests/ui/parser/f16-f128.rs (hunk 2 above)
Comment (tgross35) [reply tests/ui/parser/f16-f128.rs:1]: There didn't seem to be a good folder :) I can move it if there is a better place

I know it does nothing but I just wanted a test that can be updated as the steps move along

File: compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs (hunk 4 above)
Comment (tgross35) [reply compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:476]: There is some more discussion for these at https://github.com/rust-lang/rust/pull/114607#discussion_r1309679811 and https://github.com/rust-lang/rustc-demangle/pull/64 for this

File: compiler/rustc_smir/src/rustc_internal/internal.rs (hunk 3 above)
Comment (tgross35) [reply compiler/rustc_smir/src/rustc_internal/internal.rs]: Sure, set these to `unimplemented!`

File: compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:476]: Okay

File: compiler/stable_mir/src/ty.rs [hunk 5]
```diff
@@ -529,8 +529,10 @@ impl UintTy {
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub enum FloatTy {
+    F16,
```
Comment (compiler-errors) [inline compiler/stable_mir/src/ty.rs:532]: Pls remove

Comment (tgross35) [reply compiler/stable_mir/src/ty.rs]: Whoops yes gone

File: src/tools/clippy/clippy_utils/src/consts.rs (hunk 1 above)
Comment (tgross35) [reply src/tools/clippy/clippy_utils/src/consts.rs:781]: I added a todo list that mentions these at https://github.com/rust-lang/rust/issues/116909#issuecomment-1969554030

File: compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs (hunk 4 above)
Comment (rcvalle) [reply compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:476]: I replied on the other issue, but will reply it here as well for easier reference. My recommendation is:

* For `f128`, if it's compatible with `__float128` regardless of the underlying representation (i.e., changes between IEEE 754R 128-bit floating point type or IBM extended double depending on the target), we should always encode it as 'g' as it will be how it will also be encoded by Clang. (FWIW, it seems Clang uses IEEE 754R 128-bit floating point type instead of IBM extended double for `__float128` by default for ppc64le since May 2023 and `__ibm128` also doesn't seem to have a distinct encoding specified in the Itanium C++ ABI mangling.)
* For `f16`, if it's compatible with IEEE 754R 128-bit floating point type (`__fp16` and `_Float16` in Clang), we should always encode it as 'Dh'; if it's compatible with bfloat16 (which is unlikely?), we should always encode it as 'DF16b'. 
* The 'Dd,  'De', and 'Df' encodings are for when the Rust compiler and Clang support decimal floating point types.

tl;dr.: We should always encode `f128` as 'g', and `f16` as 'Dh'.

Comment (tgross35) [reply compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:476]: Thanks, I will update these in the upcoming PR to add intrinsics.

Total miss that they didn't make `__float128` mangle as `q` to be consistent with `f` and `d` :)

Comment (rcvalle) [reply compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:476]: Thank you! :)

**Comment by tgross35** [comment]
Yup, need somebody else's help here 😆

**Comment by compiler-errors** [comment]
@bors r+

I believe the PR was already in the queue last time I checked this morning. This PR was not given a priority, and rollup=never PRs are merged in order of the date the PR was opened, which is why the current one is merging.

**Comment by tgross35** [comment]
> @bors r+
> 
> I believe the PR was already in the queue last time I checked this morning. This PR was not given a priority, and rollup=never PRs are merged in order of the date the PR was opened, which is why the current one is merging.

🤦‍♂️ I thought it was approval date not opening date

All the same something is weird. Every time https://github.com/rust-lang/rust/pull/121242 got to the top of the queue in the past 24 hours Bors would give up and stop merging things, this one was right behind it

**Comment by tgross35** [comment]
Same PR at the top, again Bors stopped running autos. Weird

image
